design-documents/checkout-service.md

```diff
@@ -0,0 +1,123 @@
+* Totals may be calculated after adding product to cart in a separate operation
+* Totals should not be recalculated on each cart load page, only when cart changes. It will should be recalculated when placing the order to make sure product data didn't change.
+* Create service for totals calculation to reduce number of calls to BFF
+* Group calls to the same service (request all needed data in the beginning); can lead to shared state
+
+After we applying the following optimizations, we will have communication with checkout service as shown below
+* Pass customer as an argument of add to cart operation
+* Aggregate calls to catalog and inventory services Magento\CatalogInventory\Api\StockRegistryInterface::getStockItemBySku, Magento\CatalogInventory\Api\StockRegistryInterface::getStockStatusBySku need to combine into one call
+* Get product info and get final price can be combined into one call
+* Prepare item for cart logic should be eliminated, service should receive cart ready to add
+* Create API that would allow to calculate totals for the quote
+
+![First iteration](checkout-service/add-to-cart-first-iteration.png)
+
+
+If we apply these additional optimizations, communication between services will look like this
+* Have tax service return aggregated information that later can be used to display tax on product page or totals. Question: wouldn't this data be redundant in some cases?
+* Move totals in it's own service: 1) reduce number of calls to monolith, 2) cache of requested data from tax, quote and sales service and as a result reduce amount of communication.
+
+
+![Apply more decomposition](checkout-service/add-to-cart-apply-more-decomposition.png)
+
+Cons
+* More work for initial iteration as we would have to decouple more services with checkout
+
+
+Pros
+* Allows reduce calls to BFF
+* Allows to avoid intermediate state of calculating totals in BFF and reduce amount of backwards incompatible changes
+
+### Place Order Operation
+
+Communication for place order operation looks similar. The only major differences are that we need to save payment, load quote multiple times and save quote. This flow potentially can be optimized.
+
+
+![Apply more decomposition](checkout-service/place-order.png)
+
+
+## Transactions
+Currently Quote module is responsible for creating an order. Need to move this responsibility to order service (BFF after first iteration).
+
+BFF will be responsible for loading quote and managing the transaction. If order placing failed, no additional transactions need to be rolled back after first iteration as all transactions being performed on BFF. If order placing successful, but  removing quote failed, we would need to retry removing quote instead of rolling back. Mechanism of retrying of failed operations is a general concern and should be discussed separately.
+
+
+## Reporting
+For reporting purposes, service should data to reporting service (BFF after first iteration).
```

Comment:
Also, a separate generic question.

"should data" - "should send data"?

Reply:
Fixed the sentence and moved to [separate document](https://github.com/magento/architecture/pull/60).